// @fortawesome/vue-fontawesome/index.es.js
import { parse as O, icon as $, config as T, text as E } from "../../../../@fortawesome_fontawesome-svg-core@6.7.2/node_modules/@fortawesome/fontawesome-svg-core/index.js";
import { defineComponent as j, computed as d, watch as R, h as C } from "vue";
function I(e, t) {
  var r = Object.keys(e);
  if (Object.getOwnPropertySymbols) {
    var n = Object.getOwnPropertySymbols(e);
    t && (n = n.filter(function(i) {
      return Object.getOwnPropertyDescriptor(e, i).enumerable;
    })), r.push.apply(r, n);
  }
  return r;
}
function p(e) {
  for (var t = 1; t < arguments.length; t++) {
    var r = arguments[t] != null ? arguments[t] : {};
    t % 2 ? I(Object(r), !0).forEach(function(n) {
      f(e, n, r[n]);
    }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(r)) : I(Object(r)).forEach(function(n) {
      Object.defineProperty(e, n, Object.getOwnPropertyDescriptor(r, n));
    });
  }
  return e;
}
function U(e, t) {
  if (typeof e != "object" || !e) return e;
  var r = e[Symbol.toPrimitive];
  if (r !== void 0) {
    var n = r.call(e, t);
    if (typeof n != "object") return n;
    throw new TypeError("@@toPrimitive must return a primitive value.");
  }
  return (t === "string" ? String : Number)(e);
}
function q(e) {
  var t = U(e, "string");
  return typeof t == "symbol" ? t : t + "";
}
function x(e) {
  "@babel/helpers - typeof";
  return x = typeof Symbol == "function" && typeof Symbol.iterator == "symbol" ? function(t) {
    return typeof t;
  } : function(t) {
    return t && typeof Symbol == "function" && t.constructor === Symbol && t !== Symbol.prototype ? "symbol" : typeof t;
  }, x(e);
}
function f(e, t, r) {
  return t = q(t), t in e ? Object.defineProperty(e, t, {
    value: r,
    enumerable: !0,
    configurable: !0,
    writable: !0
  }) : e[t] = r, e;
}
function G(e, t) {
  if (e == null) return {};
  var r = {};
  for (var n in e)
    if (Object.prototype.hasOwnProperty.call(e, n)) {
      if (t.indexOf(n) >= 0) continue;
      r[n] = e[n];
    }
  return r;
}
function H(e, t) {
  if (e == null) return {};
  var r = G(e, t), n, i;
  if (Object.getOwnPropertySymbols) {
    var u = Object.getOwnPropertySymbols(e);
    for (i = 0; i < u.length; i++)
      n = u[i], !(t.indexOf(n) >= 0) && Object.prototype.propertyIsEnumerable.call(e, n) && (r[n] = e[n]);
  }
  return r;
}
function A(e) {
  return M(e) || V(e) || Z(e) || J();
}
function M(e) {
  if (Array.isArray(e)) return _(e);
}
function V(e) {
  if (typeof Symbol < "u" && e[Symbol.iterator] != null || e["@@iterator"] != null) return Array.from(e);
}
function Z(e, t) {
  if (e) {
    if (typeof e == "string") return _(e, t);
    var r = Object.prototype.toString.call(e).slice(8, -1);
    if (r === "Object" && e.constructor && (r = e.constructor.name), r === "Map" || r === "Set") return Array.from(e);
    if (r === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(r)) return _(e, t);
  }
}
function _(e, t) {
  (t == null || t > e.length) && (t = e.length);
  for (var r = 0, n = new Array(t); r < t; r++) n[r] = e[r];
  return n;
}
function J() {
  throw new TypeError(`Invalid attempt to spread non-iterable instance.
In order to be iterable, non-array objects must have a [Symbol.iterator]() method.`);
}
var Q = typeof globalThis < "u" ? globalThis : typeof window < "u" ? window : typeof global < "u" ? global : typeof self < "u" ? self : {}, W = { exports: {} };
(function(e) {
  (function(t) {
    var r = function(a, l, y) {
      if (!v(l) || K(l) || F(l) || k(l) || o(l))
        return l;
      var b, g = 0, B = 0;
      if (c(l))
        for (b = [], B = l.length; g < B; g++)
          b.push(r(a, l[g], y));
      else {
        b = {};
        for (var w in l)
          Object.prototype.hasOwnProperty.call(l, w) && (b[a(w, y)] = r(a, l[w], y));
      }
      return b;
    }, n = function(a, l) {
      l = l || {};
      var y = l.separator || "_", b = l.split || /(?=[A-Z])/;
      return a.split(b).join(y);
    }, i = function(a) {
      return L(a) ? a : (a = a.replace(/[\-_\s]+(.)?/g, function(l, y) {
        return y ? y.toUpperCase() : "";
      }), a.substr(0, 1).toLowerCase() + a.substr(1));
    }, u = function(a) {
      var l = i(a);
      return l.substr(0, 1).toUpperCase() + l.substr(1);
    }, m = function(a, l) {
      return n(a, l).toLowerCase();
    }, s = Object.prototype.toString, o = function(a) {
      return typeof a == "function";
    }, v = function(a) {
      return a === Object(a);
    }, c = function(a) {
      return s.call(a) == "[object Array]";
    }, K = function(a) {
      return s.call(a) == "[object Date]";
    }, F = function(a) {
      return s.call(a) == "[object RegExp]";
    }, k = function(a) {
      return s.call(a) == "[object Boolean]";
    }, L = function(a) {
      return a = a - 0, a === a;
    }, S = function(a, l) {
      var y = l && "process" in l ? l.process : l;
      return typeof y != "function" ? a : function(b, g) {
        return y(b, a, g);
      };
    }, z = {
      camelize: i,
      decamelize: m,
      pascalize: u,
      depascalize: m,
      camelizeKeys: function(a, l) {
        return r(S(i, l), a);
      },
      decamelizeKeys: function(a, l) {
        return r(S(m, l), a, l);
      },
      pascalizeKeys: function(a, l) {
        return r(S(u, l), a);
      },
      depascalizeKeys: function() {
        return this.decamelizeKeys.apply(this, arguments);
      }
    };
    e.exports ? e.exports = z : t.humps = z;
  })(Q);
})(W);
var X = W.exports, Y = ["class", "style"];
function ee(e) {
  return e.split(";").map(function(t) {
    return t.trim();
  }).filter(function(t) {
    return t;
  }).reduce(function(t, r) {
    var n = r.indexOf(":"), i = X.camelize(r.slice(0, n)), u = r.slice(n + 1).trim();
    return t[i] = u, t;
  }, {});
}
function te(e) {
  return e.split(/\s+/).reduce(function(t, r) {
    return t[r] = !0, t;
  }, {});
}
function P(e) {
  var t = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : {}, r = arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : {};
  if (typeof e == "string")
    return e;
  var n = (e.children || []).map(function(o) {
    return P(o);
  }), i = Object.keys(e.attributes || {}).reduce(function(o, v) {
    var c = e.attributes[v];
    switch (v) {
      case "class":
        o.class = te(c);
        break;
      case "style":
        o.style = ee(c);
        break;
      default:
        o.attrs[v] = c;
    }
    return o;
  }, {
    attrs: {},
    class: {},
    style: {}
  });
  r.class;
  var u = r.style, m = u === void 0 ? {} : u, s = H(r, Y);
  return C(e.tag, p(p(p({}, t), {}, {
    class: i.class,
    style: p(p({}, i.style), m)
  }, i.attrs), s), n);
}
var D = !1;
try {
  D = process.env.NODE_ENV === "production";
} catch {
}
function re() {
  if (!D && console && typeof console.error == "function") {
    var e;
    (e = console).error.apply(e, arguments);
  }
}
function h(e, t) {
  return Array.isArray(t) && t.length > 0 || !Array.isArray(t) && t ? f({}, e, t) : {};
}
function ne(e) {
  var t, r = (t = {
    "fa-spin": e.spin,
    "fa-pulse": e.pulse,
    "fa-fw": e.fixedWidth,
    "fa-border": e.border,
    "fa-li": e.listItem,
    "fa-inverse": e.inverse,
    "fa-flip": e.flip === !0,
    "fa-flip-horizontal": e.flip === "horizontal" || e.flip === "both",
    "fa-flip-vertical": e.flip === "vertical" || e.flip === "both"
  }, f(f(f(f(f(f(f(f(f(f(t, "fa-".concat(e.size), e.size !== null), "fa-rotate-".concat(e.rotation), e.rotation !== null), "fa-pull-".concat(e.pull), e.pull !== null), "fa-swap-opacity", e.swapOpacity), "fa-bounce", e.bounce), "fa-shake", e.shake), "fa-beat", e.beat), "fa-fade", e.fade), "fa-beat-fade", e.beatFade), "fa-flash", e.flash), f(f(t, "fa-spin-pulse", e.spinPulse), "fa-spin-reverse", e.spinReverse));
  return Object.keys(r).map(function(n) {
    return r[n] ? n : null;
  }).filter(function(n) {
    return n;
  });
}
function N(e) {
  if (e && x(e) === "object" && e.prefix && e.iconName && e.icon)
    return e;
  if (O.icon)
    return O.icon(e);
  if (e === null)
    return null;
  if (x(e) === "object" && e.prefix && e.iconName)
    return e;
  if (Array.isArray(e) && e.length === 2)
    return {
      prefix: e[0],
      iconName: e[1]
    };
  if (typeof e == "string")
    return {
      prefix: "fas",
      iconName: e
    };
}
var ie = j({
  name: "FontAwesomeIcon",
  props: {
    border: {
      type: Boolean,
      default: !1
    },
    fixedWidth: {
      type: Boolean,
      default: !1
    },
    flip: {
      type: [Boolean, String],
      default: !1,
      validator: function(t) {
        return [!0, !1, "horizontal", "vertical", "both"].indexOf(t) > -1;
      }
    },
    icon: {
      type: [Object, Array, String],
      required: !0
    },
    mask: {
      type: [Object, Array, String],
      default: null
    },
    maskId: {
      type: String,
      default: null
    },
    listItem: {
      type: Boolean,
      default: !1
    },
    pull: {
      type: String,
      default: null,
      validator: function(t) {
        return ["right", "left"].indexOf(t) > -1;
      }
    },
    pulse: {
      type: Boolean,
      default: !1
    },
    rotation: {
      type: [String, Number],
      default: null,
      validator: function(t) {
        return [90, 180, 270].indexOf(Number.parseInt(t, 10)) > -1;
      }
    },
    swapOpacity: {
      type: Boolean,
      default: !1
    },
    size: {
      type: String,
      default: null,
      validator: function(t) {
        return ["2xs", "xs", "sm", "lg", "xl", "2xl", "1x", "2x", "3x", "4x", "5x", "6x", "7x", "8x", "9x", "10x"].indexOf(t) > -1;
      }
    },
    spin: {
      type: Boolean,
      default: !1
    },
    transform: {
      type: [String, Object],
      default: null
    },
    symbol: {
      type: [Boolean, String],
      default: !1
    },
    title: {
      type: String,
      default: null
    },
    titleId: {
      type: String,
      default: null
    },
    inverse: {
      type: Boolean,
      default: !1
    },
    bounce: {
      type: Boolean,
      default: !1
    },
    shake: {
      type: Boolean,
      default: !1
    },
    beat: {
      type: Boolean,
      default: !1
    },
    fade: {
      type: Boolean,
      default: !1
    },
    beatFade: {
      type: Boolean,
      default: !1
    },
    flash: {
      type: Boolean,
      default: !1
    },
    spinPulse: {
      type: Boolean,
      default: !1
    },
    spinReverse: {
      type: Boolean,
      default: !1
    }
  },
  setup: function(t, r) {
    var n = r.attrs, i = d(function() {
      return N(t.icon);
    }), u = d(function() {
      return h("classes", ne(t));
    }), m = d(function() {
      return h("transform", typeof t.transform == "string" ? O.transform(t.transform) : t.transform);
    }), s = d(function() {
      return h("mask", N(t.mask));
    }), o = d(function() {
      return $(i.value, p(p(p(p({}, u.value), m.value), s.value), {}, {
        symbol: t.symbol,
        title: t.title,
        titleId: t.titleId,
        maskId: t.maskId
      }));
    });
    R(o, function(c) {
      if (!c)
        return re("Could not find one or more icon(s)", i.value, s.value);
    }, {
      immediate: !0
    });
    var v = d(function() {
      return o.value ? P(o.value.abstract[0], {}, n) : null;
    });
    return function() {
      return v.value;
    };
  }
});
j({
  name: "FontAwesomeLayers",
  props: {
    fixedWidth: {
      type: Boolean,
      default: !1
    }
  },
  setup: function(t, r) {
    var n = r.slots, i = T.familyPrefix, u = d(function() {
      return ["".concat(i, "-layers")].concat(A(t.fixedWidth ? ["".concat(i, "-fw")] : []));
    });
    return function() {
      return C("div", {
        class: u.value
      }, n.default ? n.default() : []);
    };
  }
});
j({
  name: "FontAwesomeLayersText",
  props: {
    value: {
      type: [String, Number],
      default: ""
    },
    transform: {
      type: [String, Object],
      default: null
    },
    counter: {
      type: Boolean,
      default: !1
    },
    position: {
      type: String,
      default: null,
      validator: function(t) {
        return ["bottom-left", "bottom-right", "top-left", "top-right"].indexOf(t) > -1;
      }
    }
  },
  setup: function(t, r) {
    var n = r.attrs, i = T.familyPrefix, u = d(function() {
      return h("classes", [].concat(A(t.counter ? ["".concat(i, "-layers-counter")] : []), A(t.position ? ["".concat(i, "-layers-").concat(t.position)] : [])));
    }), m = d(function() {
      return h("transform", typeof t.transform == "string" ? O.transform(t.transform) : t.transform);
    }), s = d(function() {
      var v = E(t.value.toString(), p(p({}, m.value), u.value)), c = v.abstract;
      return t.counter && (c[0].attributes.class = c[0].attributes.class.replace("fa-layers-text", "")), c[0];
    }), o = d(function() {
      return P(s.value, {}, n);
    });
    return function() {
      return o.value;
    };
  }
});
export {
  ie as FontAwesomeIcon
};
